api-samples/downloads/download_manager/popup.js

```diff
@@ -0,0 +1,775 @@
+}
+
+function getTextWidth(s) {
+  const probe = document.getElementById('text-width-probe');
+  probe.innerText = s;
+  return probe.offsetWidth;
+}
+
+function formatDateTime(date) {
+  const now = new Date();
+  const zpad_mins =
+    ':' + (date.getMinutes() < 10 ? '0' : '') + date.getMinutes();
+  if (date.getYear() != now.getYear()) {
+    return '' + (1900 + date.getYear());
+  } else if (
+    date.getMonth() != now.getMonth() ||
+    date.getDate() != now.getDate()
+  ) {
+    return (
+      date.getDate() +
+      ' ' +
+      chrome.i18n.getMessage('month' + date.getMonth() + 'abbr')
+    );
+  } else if (date.getHours() == 12) {
+    return '12' + zpad_mins + 'pm';
+  } else if (date.getHours() > 12) {
+    return date.getHours() - 12 + zpad_mins + 'pm';
+  }
+  return date.getHours() + zpad_mins + 'am';
+}
+
+function formatBytes(n) {
+  if (n < 1024) {
+    return n + 'B';
+  }
+  const prefixes = 'KMGTPEZY';
+  let mul = 1024;
+  for (let i = 0; i < prefixes.length; ++i) {
+    if (n < 1024 * mul) {
+      return (
+        parseInt(n / mul) +
+        '.' +
+        parseInt(10 * ((n / mul) % 1)) +
+        prefixes[i] +
+        'B'
+      );
+    }
+    mul *= 1024;
+  }
+  return '!!!';
+}
+
+function formatTimeLeft(ms) {
+  const prefix = 'timeLeft';
+  if (ms < 1000) {
+    return chrome.i18n.getMessage(prefix + 'Finishing');
+  }
+  const days = parseInt(ms / (24 * 60 * 60 * 1000));
+  const hours = parseInt(ms / (60 * 60 * 1000)) % 24;
+  if (days) {
+    return chrome.i18n.getMessage(prefix + 'Days', [days, hours]);
+  }
+  const minutes = parseInt(ms / (60 * 1000)) % 60;
+  if (hours) {
+    return chrome.i18n.getMessage(prefix + 'Hours', [hours, minutes]);
+  }
+  const seconds = parseInt(ms / 1000) % 60;
+  if (minutes) {
+    return chrome.i18n.getMessage(prefix + 'Minutes', [minutes, seconds]);
+  }
+  return chrome.i18n.getMessage(prefix + 'Seconds', [seconds]);
+}
+
+function ratchetWidth(w) {
+  const current = parseInt(document.body.style.minWidth) || 0;
+  document.body.style.minWidth = Math.max(w, current) + 'px';
+}
+
+function ratchetHeight(h) {
+  const current = parseInt(document.body.style.minHeight) || 0;
+  document.body.style.minHeight = Math.max(h, current) + 'px';
+}
+
+function binarySearch(array, target, cmp) {
+  let low = 0,
+    high = array.length - 1,
+    i,
+    comparison;
+  while (low <= high) {
+    i = (low + high) >> 1;
+    comparison = cmp(target, array[i]);
+    if (comparison < 0) {
+      low = i + 1;
+    } else if (comparison > 0) {
+      high = i - 1;
+    } else {
+      return i;
+    }
+  }
+  return i;
+}
+
+function arrayFrom(seq) {
+  return Array.prototype.slice.apply(seq);
+}
+
+class DownloadItem {
+  constructor(data) {
+    let item = this;
+    for (let prop in data) {
+      item[prop] = data[prop];
+    }
+    item.startTime = new Date(item.startTime);
+    if (item.canResume == undefined) {
+      DownloadItem.canResumeHack = true;
```

Comment:
This seems... like a hack 🤣  How hard do you think it would be to remove?

Reply:
lol thinking of getting

Reply:
> This seems... like a hack 🤣  How hard do you think it would be to remove?